ops/storage.py

```diff
@@ -0,0 +1,293 @@
+        self._backend.set(self.NOTICE_KEY, notices)
+
+
+class _SimpleLoader(getattr(yaml, 'CSafeLoader', yaml.SafeLoader)):
+    """Handle a couple basic python types.
+
+    yaml.SafeLoader can handle all the basic int/float/dict/set/etc that we want. The only one
+    that it *doesn't* handle is tuples. We don't want to support arbitrary types, so we just
+    subclass SafeLoader and add tuples back in.
+    """
+    # Taken from the example at:
+    # https://stackoverflow.com/questions/9169025/how-can-i-add-a-python-tuple-to-a-yaml-file-using-pyyaml
+
+
+_SimpleLoader.construct_python_tuple = yaml.Loader.construct_python_tuple
+_SimpleLoader.add_constructor(
+    u'tag:yaml.org,2002:python/tuple',
+    _SimpleLoader.construct_python_tuple)
+
+
+class _SimpleDumper(getattr(yaml, 'CSafeDumper', yaml.SafeDumper)):
+    """Add types supported by 'marshal'
+
+    YAML can support arbitrary types, but that is generally considered unsafe (like pickle). So
+    we want to only support dumping out types that are safe to load.
+    """
+
+
+_SimpleDumper.represent_tuple = yaml.Dumper.represent_tuple
+_SimpleDumper.add_representer(tuple, _SimpleDumper.represent_tuple)
+
+
+class _JujuStorageBackend:
+    """Implements the interface from the Operator framework to Juju's state-get/set/etc."""
+
+    @staticmethod
+    def is_available() -> bool:
+        """Check if Juju state storage is available.
+
+        This checks if there is a 'state-get' executable in PATH.
+        """
+        p = shutil.which('state-get')
+        return p is not None
+
+    def set(self, key: str, value: typing.Any) -> None:
+        """Set a key to a given value.
+
+        Args:
+            key: The string key that will be used to find the value later
+            value: Arbitrary content that will be returned by get().
+        Raises:
+            CalledProcessError: if 'state-set' returns an error code.
+        """
+        # default_flow_style=None means that it can use Block for
+        # complex types (types that have nested types) but use flow
+        # for simple types (like an array). Not all versions of PyYAML
+        # have the same default style.
+        encoded_value = yaml.dump(value, Dumper=_SimpleDumper, default_flow_style=None)
+        content = yaml.dump(
+            {key: encoded_value}, encoding='utf-8', default_style='|',
+            default_flow_style=False,
+            Dumper=_SimpleDumper)
+        p = subprocess.run(["state-set", "--file", "-"], input=content)
+        p.check_returncode()
+
+    def get(self, key: str) -> typing.Any:
+        """Get the bytes value associated with a given key.
+
+        Args:
+            key: The string key that will be used to find the value
+        Raises:
+            CalledProcessError: if 'state-get' returns an error code.
+        """
+        p = subprocess.run(
+            ["state-get", key],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
```

Comment:
Why are you capturing `stderr` here if not using it? note that if you want to capture both stdout and stderr in one place, you need to do `stdout=subprocess.PIPE, stderr=subprocess.STDOUT`.

Reply:
I think we just want to avoid capturing stderr. If we capture it, it doesn't end up in CalledProcessError in a clean way. If we capture it to STDOUT then it likely breaks yaml decoding.
If we don't capture it, then it ends up on the real stderr, and then it should end up in debug logs.
At least it is somewhere we can look for it.

Reply:
Ah, gotcha! Please add a comment about this in the code... thanks!

Reply:
:+1: